File: 21_Calendar/python/calendar.py
########################################################
# Calendar
#
# From: BASIC Computer Games (1978)
#       Edited by David Ahl#
#
#    This program prints out a calendar
# for any year. You must specify the
# starting day of the week of the year in
# statement 130. (Sunday(0), Monday
# (-1), Tuesday(-2), etc.) You can determine
# this by using the program WEEKDAY.
# You must also make two changes
# for leap years in statement 360 and 620.
# The program listing describes the necessary
# changes. Running the program produces a
# nice 12-month calendar.
#    The program was written by Geofrey
# Chase of the Abbey, Portsmouth, Rhode Island.
#
########################################################

from typing import Tuple


def parse_input() -> Tuple[int, bool]:
    """
    function to parse input for weekday and leap year boolean
    """

    days_mapping = {
        "sunday": 0,
        "monday": -1,
        "tuesday": -2,
        "wednesday": -3,
        "thursday": -4,
        "friday": -5,
        "saturday": -6,
    }

    day = 0
    leap_day = False

    correct_day_input = False
    while not correct_day_input:
        weekday = input("INSERT THE STARTING DAY OF THE WEEK OF THE YEAR:")

        for day_k in days_mapping.keys():
            if weekday.lower() in day_k:
                day = days_mapping[day_k]
                correct_day_input = True
                break

    while True:
        leap = input("IS IT A LEAP YEAR?:")

        if "y" in leap.lower():
            leap_day = True
            break

        if "n" in leap.lower():
            leap_day = False
            break

    return day, leap_day


def calendar(weekday, leap_year):
    """
    function to print a year's calendar.

    input:
        _weekday_: int - the initial day of the week (0=SUN, -1=MON, -2=TUES...)
        _leap_year_: bool - indicates if the year is a leap year
    """
    months_days = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    days = "S        M        T        W        T        F        S\n"
    sep = "*" * 59
    years_day = 365
    d = weekday

    if leap_year:
        months_days[2] = 29
        years_day = 366

    months_names = [
        " JANUARY ",
        " FEBRUARY",
        "  MARCH  ",
        "  APRIL  ",
        "   MAY   ",
        "   JUNE  ",
        "   JULY  ",
        "  AUGUST ",
        "SEPTEMBER",
        " OCTOBER ",
        " NOVEMBER",
        " DECEMBER",
    ]

    days_count = 0  # S in the original program

    # main loop
    for n in range(1, 13):
        days_count += months_days[n - 1]
        print(
            "** {} ****************** {} ****************** {} **\n".format(
                days_count, months_names[n - 1], years_day - days_count
            )
        )
        print(days)
        print(sep)

        for _ in range(1, 7):
            print("\n")
            for g in range(1, 8):  # noqa
                d += 1
                d2 = d - days_count

                if d2 > months_days[n]:
                    break

                if d2 <= 0:
                    print("{}".format("  "), end="       ")
                elif d2 < 10:
                    print(f" {d2}", end="       ")
                else:
                    print(f"{d2}", end="       ")
            print()

            if d2 >= months_days[n]:
                break

        if d2 > months_days[n]:
            d -= g

        print("\n")

    print("\n")


def main() -> None:
    print(" " * 32 + "CALENDAR")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print("\n" * 11)

    day, leap_year = parse_input()
    calendar(day, leap_year)


if __name__ == "__main__":
    main()

########################################################
#
########################################################
#
# Porting notes:
#
# It has been added an input at the beginning of the
# program so the user can specify the first day of the
# week of the year and if the year is leap or not.
#
########################################################

File: 21_Calendar/ruby/calendar.rb
class Calendar
  def initialize
    $stdout.sync = true    
  end

  def parse_input
    days_mapping = {
      "sunday": 0,
      "monday": -1,
      "tuesday": -2,
      "wednesday": -3,
      "thursday": -4,
      "friday": -5,
      "saturday": -6,
    }

    day = 0
    leap_day = false
    correct_day_input = false

    while !correct_day_input 
      print "INSERT THE STARTING DAY OF THE WEEK OF THE YEAR: "
      weekday = gets.chomp!

      days_mapping.each do |k, v|
        if k.to_s == weekday.downcase.to_s
          day = v
          correct_day_input = true
          break
        end
      end
    end

    while true
      print "IS IT A LEAP YEAR?: "
      leap = gets.chomp!
      if "y" == leap.downcase.to_s
        leap_day = true
        break
      end

      if "n" == leap.downcase.to_s
        leap_day = false
        break
      end
    end

    return day, leap_day
  end

  def calendar(weekday, leap_year)
    months_days = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    days = "S        M        T        W        T        F        S\n"
    sep = "*" * 59
    years_day = 365
    d = weekday

    if leap_year
      months_days[2] = 29
      years_day = 366
    end

    months_names = [
      " JANUARY ",
      " FEBRUARY",
      "  MARCH  ",
      "  APRIL  ",
      "   MAY   ",
      "   JUNE  ",
      "   JULY  ",
      "  AUGUST ",
      "SEPTEMBER",
      " OCTOBER ",
      " NOVEMBER",
      " DECEMBER",
    ]

    days_count = 0

    for n in (1..12) do
      days_count += months_days[n - 1]
      print "** #{days_count} ****************** #{months_names[n - 1]} ****************** #{years_day - days_count} **\n"
      print days
      print sep
      for nnn in (1..6) do
        print "\n"
        for g in (1..7) do
          d += 1
          d2 = d - days_count

          break if d2 > months_days[n]

          if d2 <= 0
            print "         "
          elsif d2 < 10
            print " #{d2}       "
          else
            print "#{d2}       "
          end
        end

        print "\n\n"

        break if d2 >= months_days[n]
      end
      d -= g if d2 > months_days[n]
      print "\n"
    end
    print("\n")
  end
end


calendar = Calendar.new
input = calendar.parse_input

calendar.calendar(input[0], input[1])

